### PR TITLE
feature(account)Generate token endpint and send mail event

### PR DESCRIPTION
**Pull request by wfajczyk**
# Description
Endpint POST `accounts/token/generate`
`      {
          "email": "test@ergonode.com",
          "url": "http://localhost/rest?token="
      }`

Issue #911 
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Unit Tests
- [x] Behat Test

# Checklist:
- [x] I have read the contribution requirements and fulfill them.
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by wiewiurdp — DISMISSED** [review]

File: module/account/src/Application/Validator/Constraints/AvailableHostConstraint.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+namespace Ergonode\Account\Application\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+
+/**
+ * @Annotation
+ */
+class AvailableHostConstraint extends Constraint
+{
+    public string $site = 'site';
```
Comment (piotrkreft) [inline module/account/src/Application/Validator/Constraints/AvailableHostConstraint.php:18]:
```suggestion
    public string $host = 'host';
```

File: module/account/src/Application/Validator/Constraints/AvailableHostConstraint.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+    public string $validMessage = 'This {{ site }} is not available';
```
Comment (piotrkreft) [inline module/account/src/Application/Validator/Constraints/AvailableHostConstraint.php:20]:
```suggestion
    public string $validMessage = 'This {{ host }} is not available';
```

File: module/account/src/Application/Form/Model/GenerateUserTokenModel.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+     */
+    public ?string $email;
+
+    /**
+     * @Assert\NotBlank()
+     * @Assert\Url()
+     *
+     * @AvailableHostConstraint(site="path")
+     */
+    public ?string $path;
```
Comment (piotrkreft) [inline module/account/src/Application/Form/Model/GenerateUserTokenModel.php:28]:
```suggestion
    public ?string $url;
```
path is just a part without host name etc

File: module/account/src/Application/Validator/Constraints/AvailableHostConstraintValidator.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+namespace Ergonode\Account\Application\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+use Webmozart\Assert\Assert;
+
+class AvailableHostConstraintValidator extends ConstraintValidator
+{
+    private array $sites;
```
Comment (piotrkreft) [inline module/account/src/Application/Validator/Constraints/AvailableHostConstraintValidator.php:18]:
```suggestion
    private array $hosts;
```

File: module/account/src/Application/DependencyInjection/Configuration.php [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    public function getConfigTreeBuilder(): TreeBuilder
+    {
+        $treeBuilder = new TreeBuilder('ergonode_account');
+        /** @phpstan-ignore-next-line */
+        $treeBuilder
+            ->getRootNode()
+                ->children()
+                    // TODO workaround for not being able to set json array from env variables
+                    // see https://github.com/symfony/symfony/issues/27683
+                    ->variableNode('host')
```
Comment (piotrkreft) [inline module/account/src/Application/DependencyInjection/Configuration.php:25]:
```suggestion
                    ->variableNode('hosts')
                    ->info('List of hosts allowed to generate like to reset password')
```

File: module/account/src/Domain/Event/User/UserResetTokenGeneratedEvent.php [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace Ergonode\Account\Domain\Event\User;
+
+use Ergonode\Account\Domain\ValueObject\ResetToken;
+use Ergonode\EventSourcing\Infrastructure\DomainEventInterface;
+use Ergonode\SharedKernel\Domain\Aggregate\UserId;
+use Ergonode\SharedKernel\Domain\AggregateId;
+use JMS\Serializer\Annotation as JMS;
+
+class UserResetTokenGeneratedEvent implements DomainEventInterface
```
Comment (piotrkreft) [inline module/account/src/Domain/Event/User/UserResetTokenGeneratedEvent.php:17]: I don't think it should be an event sourcing event. It's just confusing.

File: module/account/src/Domain/Event/User/UserResetTokenGeneratedEvent.php [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+{
+    /**
+     * @JMS\Type("Ergonode\SharedKernel\Domain\Aggregate\UserId")
```
Comment (piotrkreft) [inline module/account/src/Domain/Event/User/UserResetTokenGeneratedEvent.php:20]: Isn't the configuration redundant?

File: module/account/src/Domain/Event/User/UserResetTokenGeneratedEvent.php [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    private string $path;
+
+    public function __construct(UserId $id, ResetToken $token, string $path)
+    {
+        $this->id = $id;
+        $this->token = $token;
+        $this->path = $path;
+    }
+
+    public function getAggregateId(): AggregateId
```
Comment (piotrkreft) [inline module/account/src/Domain/Event/User/UserResetTokenGeneratedEvent.php:41]:
```suggestion
    public function getUserId(): UserId
```

File: module/account/src/Domain/Command/User/GenerateUserResetPasswordTokenCommand.php [hunk 9]
```diff
@@ -19,13 +19,24 @@ class GenerateUserResetPasswordTokenCommand implements AccountCommandInterface
      */
     private UserId $id;
 
-    public function __construct(UserId $id)
+    /**
+     * @JMS\Type("string")
+     */
+    private string $path;
+
+    public function __construct(UserId $id, string $path)
```
Comment (piotrkreft) [inline module/account/src/Domain/Command/User/GenerateUserResetPasswordTokenCommand.php:27]:
```suggestion
    public function __construct(UserId $id, string $url)
```

File: module/account/features/account-token.feature [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+  Scenario: Reset Token user
+    And I add "Content-Type" header equal to "application/json"
+    And I add "Accept" header equal to "application/json"
+    When I send a POST request to "/api/v1/accounts/token/generate" with body:
+      """
+      {
+          "email": "test@ergonode.com",
+          "path": "http://localhost/rest?token="
+      }
+      """
```
Comment (piotrkreft) [inline module/account/features/account-token.feature:12]: Is it possible to verify whether the mail has been passed to be sent?

**Review by piotrkreft — CHANGES_REQUESTED** [review]

File: module/account/features/account-token.feature (hunk 10 above)
Comment (wfajczyk) [reply module/account/features/account-token.feature:12]: I don't know, I'll check in the next CR

**Review by piotrkreft — APPROVED** [review]

File: module/account/src/Application/DependencyInjection/ErgonodeAccountExtension.php [hunk 11]
```diff
@@ -19,9 +19,10 @@
 class ErgonodeAccountExtension extends Extension implements PrependExtensionInterface
 {
     /**
+     * @throws \Exception
+     *
      * @param array $configs
      *
```
Comment (wiewiurdp) [inline module/account/src/Application/DependencyInjection/ErgonodeAccountExtension.php:25]: are this docs neccesary  ?

File: module/account/src/Application/Form/GenerateUserTokenForm.php [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class GenerateUserTokenForm extends AbstractType
+{
+    /**
+     * @param array $options
+     */
```
Comment (wiewiurdp) [inline module/account/src/Application/Form/GenerateUserTokenForm.php:21]: are this docs neccesary ?

**Review by wiewiurdp — APPROVED** [review]

**Pull request merged**